### ISSUE TITLE
@Optional doesn't work correctly in combination with @OutputFile/@OutputFiles, @OutputDirectory/@OutputDirectories

### ISSUE THREAD

**Issue by hungvietnguyen** [gradle/gradle#3073]
For incremental builds, the `@Optional` annotation works correctly when used together with `@Input` but not with `@InputFile/@InputDirectory/@OutputFile/@OutputDirectory`. In those cases, changing a property (from a null value to a non-null value and vice versa, or even from a non-null value to another non-null value), the task is still UP-TO-DATE.

Steps to reproduce:

1. Use the build.gradle file as inlined below. (Also create a 'src' directory next to the file.)
2. Run ./gradlew copy --info. Observe that the copy task is run.
3. Run ./gradlew copy --info -Plogfile=log.txt. Observe that copy task is UP-TO-DATE. This is not expected.
4. Run ./gradlew copy --info -Plogfile=log2.txt. Observe that copy task is again UP-TO-DATE. This is also not expected.
```
class CustomCopyTask extends Copy {
    File logFile

    @OutputFile
    @Optional
    File getOutputFile() {
        return logFile
    }

    @TaskAction
    void doAction() {
	if (logFile != null) {	    
	    logFile.write "Task is running"
	}
    }
}

task copy(type: CustomCopyTask) {
    from 'src'
    into 'build'
    logFile = project.hasProperty('logfile') ? new File(logfile) : null
}
```

**Comment by big-guy** [comment]
Which version of Gradle are you using?

**Comment by hungvietnguyen** [comment]
I'm using Gradle 4.1. (Sorry, should have mentioned that.)

**Comment by hungvietnguyen** [comment]
Note that this bug appears only after the property is assigned to a null value (step 2 above). Before that, @Optional works just fine: changing a non-null value to another non-null value triggers re-run. (In the above example, if step 2 is skipped, step 3 and 4 will work just fine.) After the property is set to null, then changing it from any value to any other value will never trigger a re-run.

Related bug in the Android Gradle plugin where this issue was first noticed: https://issuetracker.google.com/65829618

**Comment by lptr** [comment]
I can confirm this breaks with `@OutputFile`. `@InputFile` works fine, though:

```text
:customTask
Putting task artifact state for task ':customTask' into context took 0.0 secs.
Executing task ':customTask' (up-to-date check took 0.009 secs) due to:
  Input property 'inputFile' file /Users/lptr/Workspace/gradle/gradle/build/tmp/test files/TaskUpToDateIntegrationTest/optional_input_chan...p_to_date/8797y/input.txt has been added.
:customTask (Thread[Task worker for ':',5,main]) completed. Took 0.022 secs.
```

I've added tests here: https://github.com/gradle/gradle/commit/67c74c048a29b5b595d34354452392539c8ff3e0

**Comment by hungvietnguyen** [comment]
Yeah, I just checked again and can also confirm that @InputFile/@InputFiles works, but @OutputFile/@OutputFiles, @InputDirectory, @OutputDirectory/@OutputDirectories don't work.

**Comment by wolfs** [comment]
@hungvietnguyen The fix has been merged an will be included in Gradle 4.3.

**Comment by hungvietnguyen** [comment]
That sounds good, thanks!
